ocaml-lsp-server/src/ocamlformat.ml

```diff
@@ -0,0 +1,119 @@
+open Import
+
+module File_type = struct
+  type t =
+    | Impl
+    | Intf
+    | Name of string
+
+  let to_cmdline_args = function
+    | Impl -> [ "--impl" ]
+    | Intf -> [ "--intf" ]
+    | Name n -> [ Printf.sprintf "--name=%s" n ]
+end
+
+module Input = struct
+  type t =
+    | Stdin of string * File_type.t
+    | File of string
+end
+
+module Output = struct
+  type _ t =
+    | File : string -> unit t
+    | Stdout : string t
+
+  let to_cmdline_args : type o. o t -> string list = function
+    | File s -> [ Printf.sprintf "--output=%s" s ]
+    | Stdout -> []
+end
+
+module Options = struct
+  type t = string list
+
+  let default : t = []
+end
+
+type 'result command =
+  | FormatFile : Input.t * 'result Output.t -> 'result command
+  | FormatFilesInPlace : string list -> unit command
+  | Check : Input.t -> bool command
+
+let read_to_end ?(hint = 0) (inChan : in_channel) : string =
+  let buf = Buffer.create hint in
+  let rec go () =
+    try
+      let line = input_line inChan in
+      Buffer.add_string buf line;
+      Buffer.add_char buf '\n';
+      go ()
```

Comment:
Note that this implementation isn't tail recursive.

You should use `Io.read_all` in any case however.

Reply:
If I understand correctly that function exists in ExtLib and we don't use it as of now. Is it OK to add it to the dependencies?

Reply:
No, it exists in stdune in fact. So it should already be available to you.

Reply:
Nice! It would be great to have something like Hoogle, though.